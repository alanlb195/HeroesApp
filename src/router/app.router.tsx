import { createBrowserRouter, Navigate } from "react-router";
import { HomePage } from "@/heroes/pages/home/HomePage";
import { HeroPage } from "@/heroes/pages/hero/HeroePage";
// import { SearchPage } from "@/heroes/pages/search/SearchPage";
import { AdminPage } from "@/admin/pages/AdminPage";
import { HeroesLayout } from "@/heroes/layouts/HeroesLayout";
import { AdminLayout } from "@/admin/layouts/AdminLayout";
import { lazy } from "react";

const SearchPage = lazy(() => import("@/heroes/pages/search/SearchPage"));

export const appRouter = createBrowserRouter([

  // Heroes
  {
    path: "/",
    element: <HeroesLayout />,
    children: [
      {
        index: true,
        element: <HomePage />,
      },
      {
        path: "hero/:slugId",
        element: <HeroPage />,
      },
      {
        path: "search",
        element: <SearchPage />,
      },
    ]
  },

  //Admin
  {
    path: "/admin",
    element: <AdminLayout />,
    children: [
      {
        path: "/admin",
        element: <AdminPage />,
      },
    ]
  },

  // Not Found
  {
    path: "*",
    element: <Navigate to="/" />
  },
]);