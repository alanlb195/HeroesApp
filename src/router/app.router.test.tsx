import { describe, expect, test, vi } from 'vitest';
import { appRouter } from './app.router';
import { render, screen } from '@testing-library/react';
import { createMemoryRouter, Outlet, RouterProvider, useParams } from 'react-router';


vi.mock("@/heroes/layouts/HeroesLayout", () => ({
    HeroesLayout: () => <div data-testid="heroes-layout"><Outlet /></div>
}));

vi.mock("@/heroes/pages/home/HomePage", () => ({
    HomePage: () => <div data-testid="home-page"></div>
}));

vi.mock("@/heroes/pages/search/SearchPage", () => ({
    default: () => <div data-testid="search-page"></div>
}))

vi.mock("@/heroes/pages/hero/HeroePage", () => ({
    HeroPage: () => {
        const { slugId = '' } = useParams();
        return (
            <div data-testid="hero-page">
                HeroPage - {slugId}
            </div>
        )
    }
}));


describe('appRouter', () => {
    test('should be configured as expected', () => {
        expect(appRouter.routes).toMatchSnapshot();
    });

    test('should render home page at root path', async () => {

        const router = createMemoryRouter(appRouter.routes, {
            initialEntries: ['/']
        });

        render(<RouterProvider router={router} />)
        // screen.debug();
        expect(screen.getByTestId('home-page')).toBeDefined();
    });


    test('should render hero page at /heroes/:idSlug path', async () => {

        const router = createMemoryRouter(appRouter.routes, {
            initialEntries: ['/hero/superman'],
        });

        render(<RouterProvider router={router} />)

        // screen.debug();
        expect(screen.getByTestId('hero-page').innerHTML).toContain('superman');
    });

    test('should render search page at /search path', async () => {
        const router = createMemoryRouter(appRouter.routes, {
            initialEntries: ['/search'],
        });

        render(<RouterProvider router={router} />);
        expect(await screen.findByTestId('search-page')).toBeDefined();
        // screen.debug();
    });

    test('should redirect to home for unknown routes', async () => {
        const router = createMemoryRouter(appRouter.routes, {
            initialEntries: ['/another-page-404'],
        });

        render(<RouterProvider router={router} />);
        // screen.debug();
        expect(screen.getByTestId('home-page')).toBeDefined();
    });

});
