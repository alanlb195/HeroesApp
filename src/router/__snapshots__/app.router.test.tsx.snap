// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`appRouter > should be configured as expected 1`] = `
[
  {
    "children": [
      {
        "element": <HomePage />,
        "hasErrorBoundary": false,
        "id": "0-0",
        "index": true,
      },
      {
        "children": undefined,
        "element": <HeroPage />,
        "hasErrorBoundary": false,
        "id": "0-1",
        "path": "hero/:slugId",
      },
      {
        "children": undefined,
        "element": <UNDEFINED />,
        "hasErrorBoundary": false,
        "id": "0-2",
        "path": "search",
      },
    ],
    "element": <HeroesLayout />,
    "hasErrorBoundary": false,
    "id": "0",
    "path": "/",
  },
  {
    "children": [
      {
        "children": undefined,
        "element": <AdminPage />,
        "hasErrorBoundary": false,
        "id": "1-0",
        "path": "/admin",
      },
    ],
    "element": <AdminLayout />,
    "hasErrorBoundary": false,
    "id": "1",
    "path": "/admin",
  },
  {
    "children": undefined,
    "element": <Navigate
      to="/"
    />,
    "hasErrorBoundary": false,
    "id": "2",
    "path": "*",
  },
]
`;
